Fix two bugs in timeSeriesStats output in MPAS-Ocean

xylar opened the pull request:
This merge fixes two issues in `timeSeriesStatsDaily` and `timeSeriesStatsMonthly` in MPAS-Ocean.

The first is that the value `xtimeMonthly_start` was inaccurate at the start of a simulation and after model restart.  This occurred because of a combination of problems.  
1. When E3SM reset the MPAS clock, MPAS alarms are recalibrated so they ring immediately (rather than at the end of the desired time-averaging interval), a behavior different from MPAS standalone runs. 
2. As a result, it turns out to be desirable to immediately check (and reset) alarms when MPAS) starts for the first time or restarts, allowing the time interval to begin afresh and time averaging to start over if appropriate.
3. Because MPAS-O is a step behind when starting a new E3SM simulation, it is desirable to compute time-series stats on startup (but not on restart) so the first day or month is complete (rather than short one time step).
4. `xtimeMonthly_start` was being computed by subtracting a time step from the current time, rather than a compute interval of `timeSeriesStats` after a restart.

This issue was present in E3SM output.
closes #5217

The second issue does not affect E3SM output because E3SM doesn't use restart files for `timeSeriesStats`.  However, standalone runs that do use restart files for this analysis were experiencing a problem in which time averaging was not getting reset after a restart, leading to time averages over the whole simulation rather than the expected (daily or monthly) interval.  This issue has been fixed by referencing the reset alarm to the `simulationStartTime` rather than the beginning of the current (possibly restart) run.

closes https://github.com/MPAS-Dev/MPAS-Model/issues/121

xylar committed:
Fix start time for time series stats interval

Before this merge, it was incorrectly assumed to be one time
step before the current time.  The correct interval depends on
the compute interval, which has now been made available in the
form of the `computeInterval`.

xylar committed:
Don't write Time and Time_bnds to AM restart files

These variables will get read back on restart into the main
system versions of these variables, which is not what we would
want.  Instead, on restart, Time_bnds(1) needs to be
recomputed from xtime_start

xylar committed:
Switch timeSeriesStats to use simulationStartTime

Previously, timeSeriesStats was using the start time of the current
run as the beginning of stats.  Since this time starts over at
restarts, it was not triggering a reset of the stats as it should
have at the beginning of a restart interval that is commensurate
with the time-averaging interval.

With this merge, we now use the `simulationStartTime` to determine
when time-averaging began, so that alarms are correctly triggered
on restart.

xylar commented:
This PR is based off of #5202 and will need to be rebased after that goes in.

xylar commented:
@jonbob, would you be willing to run a 1-year E3SM simulation with this branch that you can compare with something you already have handy?  That should be plenty to verify that ``xtimeDaily_start` and xtimeMonthly_start` is fixed and that all other variables are unchanged.  This can wait until #5202 goes in or a test run with #5202 could be the baseline, whichever is more efficient for you.

xylar commented:
@mark-petersen, I'd like this fix to go in before the code freeze.  Any chance you could review it soon?

jonbob commented:
@xylar -- I'll be happy to test, doing whatever runs you think are necessary. I'll prioritize getting #5202 in as well.

mark-petersen commented:
Thanks, I'm reviewing now. MPAS-Seaice has the identical timeSeriesStats. Does it need the identical fix?

xylar commented:
Yes, eventually, MPAS-Seaice will need this and also the equivalent of #5202.  It seems hard to get both of those in and tested before the code freeze, though.

xylar commented:
I tried an E3SM test (`ERS_D.ne11_oQU240.WCYCL1850NS.anvil_intel`) and it fails.  Let me figure that out before you do anymore review or testing.

mark-petersen reviewed (approved):
Tested with gnu debug, gnu optimized, and intel optimized using all daily and monthly timeSeriesStats tests and https://github.com/MPAS-Dev/compass/pull/435. This PR passes all. Based on @xylar's comments, it looks like this fixes the problem described.

xylar commented:
Update: `ERS_D.ne11_oQU240.WCYCL1850NS.anvil_intel` now runs with the fix I pushed. I'm testing `ERS.ne11_oQU240.WCYCL1850NS.anvil_intel` against a baseline from #5202 and will report back as soon as that finishes.

xylar commented:
`ERS.ne11_oQU240.WCYCL1850NS.anvil_intel` passes and is BFB with #5202.  @jonbob, this is ready for you to review but maybe fine to wait until #5202 goes in.

jonbob commented:
note: running this PR in E3SM did not initially achieve the expected results

xylar commented:
Yes, I'm working on debugging this.  It is proving quite challenging.

xylar committed:
Compute timeSeriesStats on startup unless restarting

With this merge, we copute timeSeriesStats on startup if we are
doing a fresh E3SM run but not if we are restarting.  This ensures
that we have a complete first day, month, etc.

xylar committed:
Update registry comments for compute_on_startup

xylar commented:
@mark-petersen and @jonbob, I have updated the PR description and requested a re-review.

xylar commented on components/mpas-ocean/bld/build-namelist:
```diff
@@ -1446,7 +1450,11 @@ add_default($nl, 'config_AM_timeSeriesStatsDaily_backward_output_offset');
 
 add_default($nl, 'config_AM_timeSeriesStatsMonthly_enable');
-add_default($nl, 'config_AM_timeSeriesStatsMonthly_compute_on_startup');
+if ($CONTINUE_RUN eq 'TRUE') {
+	add_default($nl, 'config_AM_timeSeriesStatsMonthly_compute_on_startup', 'val'=>".false.");
+} else {
+	add_default($nl, 'config_AM_timeSeriesStatsMonthly_compute_on_startup', 'val'=>".true.");
+}
```
This and similar changes above and below for other time-series stats seems to be necessary so that the first day or month in an E3SM simulation is complete.  Let me know if you have concerns.

xylar commented on components/mpas-ocean/src/analysis_members/Registry_time_series_stats_climatology.xml:
```diff
@@ -8,7 +8,7 @@
 			type="logical"
 			default_value=".false."
-			description="Logical flag determining if an analysis member computation occurs on start-up. You likely want this off for this (time series) analysis member because it will accumulate any state prior to time stepping (double counting the last time step)."
+			description="Logical flag determining if an analysis member computation occurs on start-up. This should be set to .false. for better efficiency in most cases because the results of this first call will simply be thrown away.  The exception is at the beginning of new E3SM simulations (CONTINUE_RUN = FALSE), where the first time step will be missing if this is not set to .true."
```
This and related descriptions for other time-series stats have been updated for better accuracy and clarity.  @jonbob, I didn't try to update the associate files in `bld` because I don't remember the exact process.  I can look that up, though, if you would like me to take care of it.

xylar commented:
## Re-testing

I again ran the 4 new tests introduced in https://github.com/MPAS-Dev/compass/pull/435 to show that these fixes work as expected in standalone MPAS-Ocean.

I also ran 3 months of the E3SM test case `CMPASO-JRA1p4_TL319_EC30to60E2r2` on Chrysalis, with 2 restarts in between.  I see (as expected) the following `Time` and `Time_bnds` values:

```
$ ncdump -v Time,Time_bnds 20221021.CMPASO-JRA1p4_TL319_EC30to60E2r2.fix-time-series-stats.chrysalis.mpaso.hist.am.timeSeriesStatsMonthly.0001-01-01.nc | tail

 Time = 15.5 ;

 Time_bnds =
  0, 31 ;

$ ncdump -v Time,Time_bnds 20221021.CMPASO-JRA1p4_TL319_EC30to60E2r2.fix-time-series-stats.chrysalis.mpaso.hist.am.timeSeriesStatsMonthly.0001-02-01.nc | tail

 Time = 45 ;

 Time_bnds =
  31, 59 ;
}

$ ncdump -v Time,Time_bnds 20221021.CMPASO-JRA1p4_TL319_EC30to60E2r2.fix-time-series-stats.chrysalis.mpaso.hist.am.timeSeriesStatsMonthly.0001-03-01.nc | tail

 Time = 74.5 ;

 Time_bnds =
  59, 90 ;
}
```
See results in:
```
/lcrc/group/e3sm/ac.xylar/scratch/chrys/20221021.CMPASO-JRA1p4_TL319_EC30to60E2r2.fix-time-series-stats.chrysalis/run
```

xylar commented:
Getting there but testing in MPAS-Seaice is still showing problems that I want to understand before merging this.

rljacob commented:
telecon notes:  waiting for @xylar to approve.

xylar commented:
Still debugging...

xylar commented:
I looked into this further today.  Similar fixes in MPAS-Seaice revealed a problem with this approach.  I have an alternative in mind that I will try to implement next week.  If this PR being open is a problem, we can close it for now and re-open later.

rljacob commented:
notes:  had to be redone.  Still in progress.

xylar commented:
I can close this for now as well.
